File: src/main/java/im/tox/tox4j/crypto/exceptions/ToxDecryptionException.java
package im.tox.tox4j.crypto.exceptions;

import im.tox.tox4j.annotations.NotNull;
import im.tox.tox4j.exceptions.ToxException;

public final class ToxDecryptionException extends ToxException<ToxDecryptionException.Code> {

  public enum Code {
    /**
     * The input data is missing the magic number (i.e. wasn't created by this
     * module, or is corrupted)
     */
    BAD_FORMAT,
    /**
     * The encrypted byte array could not be decrypted. Either the data was
     * corrupt or the password/key was incorrect.
     */
    FAILED,
    /**
     * The input data was shorter than {@link ToxConstants.PASS_ENCRYPTION_EXTRA_LENGTH} bytes.
     */
    INVALID_LENGTH,
    /**
     * The crypto lib was unable to derive a key from the given passphrase,
     * which is usually a lack of memory issue. The functions accepting keys
     * do not produce this error.
     */
    KEY_DERIVATION_FAILED,
    /**
     * Some input data, or maybe the output pointer, was null.
     */
    NULL,
  }

  public ToxDecryptionException(@NotNull Code code) {
    this(code, "");
  }

  public ToxDecryptionException(@NotNull Code code, String message) {
    super(code, message);
  }

}

File: projects/tox4j/src/main/java/im/tox/tox4j/crypto/exceptions/ToxDecryptionException.scala
package im.tox.tox4j.crypto.exceptions

import im.tox.tox4j.exceptions.ToxException

object ToxDecryptionException {

  sealed trait Code extends ToxException.Code
  /**
   * The input data is missing the magic number (i.e. wasn't created by this
   * module, or is corrupted)
   */
  case object BAD_FORMAT extends Code
  /**
   * The encrypted byte array could not be decrypted. Either the data was
   * corrupt or the password/key was incorrect.
   */
  case object FAILED extends Code
  /**
   * The input data was shorter than [[ToxCryptoConstants.ENCRYPTION_EXTRA_LENGTH]] bytes.
   */
  case object INVALID_LENGTH extends Code
  /**
   * The key or input data was null or empty.
   */
  case object NULL extends Code

}

final case class ToxDecryptionException(code: ToxDecryptionException.Code, message: String = "")
  extends ToxException[ToxDecryptionException.Code](message)
